vault/ha.go

```diff
@@ -408,6 +410,58 @@ func (c *Core) drainPendingRestarts() {
 	}
 }
 
+func (c *Core) runStandbyGrabStateLock(stopCh <-chan struct{}) error {
+	acquiredCh := make(chan struct{})
+	quitCh := make(chan struct{})
+
+	// We want to quit lock acquisition on two conditions:
+	//
+	// 1. Timeout.
+	// 2. If we're told to stop.
+	//
+	// While technically unnecessary, we only quit lock acquisition
+	// if we've not been canceled.
+	go func() {
+		timeout := time.NewTimer(DefaultMaxRequestDuration)
+
+		canceled := false
+		select {
+		case <-timeout.C:
+			canceled = true
+		case <-stopCh:
+			canceled = true
+		case <-acquiredCh:
+		}
+
+		if !timeout.Stop() {
+			<-timeout.C
+		}
```

Comment:
I don't get this part, why do we need this?

Reply:
Timers will leak channels if you don't stop them. When the channel reports that it is already stopped, you need to consume the value out of the channel as nobody else will.